### PR TITLE
Expand contributor docs

### PR DESCRIPTION
**Pull request by mcmire**
The monorepo works differently than other repos, and we want to make sure that contributors are able to use it on their own:

* The release process is different than other repos because it makes use of `create-release-branch` rather than the `create-release-pr` action.
* Performing tasks (running tests, etc.) across the monorepo is different than with other repos because it takes advantage of workspaces.
* Testing changes made to the monorepo in other projects is different because of the use of preview builds.

This commit moves the existing contributor documentation which was formerly present in the README to a separate file and adds some more sections.

### CODE REVIEW

File: docs/contributing.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# Contributing to the monorepo
+
+### Getting started
```
Comment (Mrtenz) [inline docs/contributing.md:3]: Nit: This (and some other headings in this file) should be `##`.

```suggestion
## Getting started
```

Comment (mcmire) [reply docs/contributing.md]: Whoops! Thank you.

File: docs/contributing.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+### Running unit tests
+
+- Run `yarn workspace <workspaceName> run test` to run all tests for a package.
+- Run `yarn workspace <workspaceName> run jest --no-coverage <file>` to run a test file within the context of a package.
+- Run `yarn test` to run tests for all packages.
+
+> **Note**  
+> `workspaceName` here is the `name` field within a package's `package.json`, e.g., `@metamask/announcement-controller`, not the directory where it is located, e.g., `packages/announcement-controller`.
+
+### Integrating with other projects
```
Comment (legobeat) [inline docs/contributing.md:22]: Overall these sections and structure are a big step up, great work!

To me this feels more like its own section rather than subtopic of Testing (so `##` rather than `###`).

Nit: Section title could still be more descriptive? Like `Use your fork as a dependency`, `Include custom builds in other projects` or some such other? I guess it's tight to not get too wordy though)

File: docs/contributing.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+When developing changes to packages within this repository that a different project depends upon, you may wish to load those changes into the project and test them locally or in CI before publishing proper releases of those packages. To solve that problem, this repository provides a mechanism to publish "preview" versions of packages to GitHub Package Registry. These versions can then be used in the project like any other version, provided the project is configured to use that registry.
+
+#### As a MetaMask engineer
```
Comment (legobeat) [inline docs/contributing.md:26]: This is a little bit fuzzy (maybe intentional? If so that's great and I think it's appropriate to keep)

Flavors:
  - If we want to put a clear line (risking being kind of bureaucratic about it): `As a MetaMask organization member`
  - Implied might work: `As a MetaMask member`
  - To make it more open: `As a MetaMask developer`
  - To make that more inviting: `As a MetaMask contributor`

File: docs/contributing.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+   @metamask:registry=https://npm.pkg.github.com
+   //npm.pkg.github.com/:_authToken=<your personal access token>
+   ```
+   Make sure not to commit this file.
+3. Go to GitHub and open up a pull request for this repository (for the extension or mobile app, this needs to be a draft PR), then post a comment on the PR with the text `@metamask-bot preview-build`. (This triggers the `publish-preview` GitHub action.)
+4. After a few minutes, you will see a new comment indicating that all packages have been published with the format `<package name>-<commit id>`.
+5. Switch back to your project locally and update `package.json` by replacing the versions for the packages you've changed in your PR using the new version format (e.g. `1.2.3-e2df9b4` instead of `~1.2.3`), then run `yarn install`.
+6. Repeat steps 3-5 after pushing new changes to your PR to generate and use new preview versions.
+
+#### As an external contributor
```
Comment (legobeat) [inline docs/contributing.md:42]: How about `As an independent developer`? Communicates the same message without the connotation of being "outside" or the necessity of "contributing" (even if that's what we hope is the result of reading this document of course ;))

File: docs/contributing.md (hunk 3 above)
Comment (mcmire) [reply docs/contributing.md:26]: Good point, I was thinking of a short way of saying the first thing. I like "As a MetaMask contributor".

File: docs/contributing.md (hunk 4 above)
Comment (mcmire) [reply docs/contributing.md]: Fair! Updated in [8770a7a](https://github.com/MetaMask/controllers/pull/993/commits/8770a7a95c57c3936c8d5412c9a806a078fd9a2f).

File: docs/contributing.md (hunk 3 above)
Comment (mcmire) [reply docs/contributing.md]: Updated in [b321b41](https://github.com/MetaMask/controllers/pull/993/commits/b321b413beb5d24ddd7cbea31e3e27bbaa93c80b).

File: docs/contributing.md (hunk 2 above)
Comment (mcmire) [reply docs/contributing.md:22]: > To me this feels more like its own section rather than subtopic of Testing

Sure. It's a distinct process so that makes sense.

> Section title could still be more descriptive

Fair. I was thinking about this in terms of people who may not work with this repo from day to day. For instance, I'm accustomed to working within the extension, but now I've switched to this repo because I've updated some controllers and I've confirmed that the changes work on their own, but I want to go back to the thing that actually uses them and make sure I haven't broken anything. What would you call that step? I was thinking along the lines of manual integration testing. Perhaps it can be thought about differently though. 

In any case, "Include custom builds in other projects" seems closer generally speaking. Maybe "Creating and using on-demand builds in other projects"? I'm not sure I like how technical that would sound, but it _is_ accurate. Hmm...

File: docs/contributing.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+> **Note**  
+> `workspaceName` here is the `name` field within a package's `package.json`, e.g., `@metamask/announcement-controller`, not the directory where it is located, e.g., `packages/announcement-controller`.
+
+### Integrating with other projects
+
+When developing changes to packages within this repository that a different project depends upon, you may wish to load those changes into the project and test them locally or in CI before publishing proper releases of those packages. To solve that problem, this repository provides a mechanism to publish "preview" versions of packages to GitHub Package Registry. These versions can then be used in the project like any other version, provided the project is configured to use that registry.
+
+#### As a MetaMask contributor
+
+If you're a MetaMask engineer, you can create these preview versions via draft pull requests:
```
Comment (Gudahtt) [inline docs/contributing.md:28]: Nit: should use the same term here presumably

```suggestion
If you're a MetaMask contributor, you can create these preview versions via draft pull requests:
```

Side-note, I've mostly become accustomed to it now, but I still hesitate to use the term "engineer" to describe developers because in Canada it's a protected term, only to be used by professional engineers. At least in theory. Plus there are non-developer contributors too.

Comment (mcmire) [reply docs/contributing.md]: Ah - good point! I'd forgotten about that.

File: docs/contributing.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+1. Navigate to your settings within GitHub and [create a **classic** access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#creating-a-personal-access-token-classic). Make sure to give this token the `packages:read` scope.
+2. Switch to your project locally and add a `.npmrc` file with the following content, filling in the appropriate areas:
+   ```
+   @<your GitHub username>:registry=https://npm.pkg.github.com
+   //npm.pkg.github.com/:_authToken=<your personal access token>
+   ```
+   Make sure not to commit this file.
+3. Open the `package.json` for each package that you want to publish and change the scope in the name from `@metamask` to `@<your GitHub username>`.
+4. Switch to your fork of this repository locally and run `commit_sha=$(git rev-parse --short HEAD)`. This generates the identifier appended to all package versions to associate them with the current commit. (Alternatively, you can set `commit_sha` to any arbitrary commit id.)
```
Comment (Gudahtt) [inline docs/contributing.md:54]: Nit: Could we skip this step, and use `yarn prepare-preview-builds "$(git rev-parse --short HEAD)"` in the next step instead? Better one command than two. I don't think we should be encouraging people to choose any arbitrary id either, that could be confusing (it will always use the local tree to publish).

I see that it is referenced later as well, but it should be easy enough for users to see what the preview build numbers are (it's in the CLI output, and users can always check the package manifests for reference).

File: docs/contributing.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+If you're a MetaMask engineer, you can create these preview versions via draft pull requests:
+
+1. Navigate to your settings within GitHub and [create a classic access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#creating-a-personal-access-token-classic). Make sure to give this token the `packages:read` scope.
+2. Switch to your project locally and add a `.npmrc` file with the following content, filling in the appropriate areas:
+   ```
+   @metamask:registry=https://npm.pkg.github.com
+   //npm.pkg.github.com/:_authToken=<your personal access token>
+   ```
+   Make sure not to commit this file.
+3. Go to GitHub and open up a pull request for this repository (for the extension or mobile app, this needs to be a draft PR), then post a comment on the PR with the text `@metamask-bot preview-build`. (This triggers the `publish-preview` GitHub action.)
```
Comment (Gudahtt) [inline docs/contributing.md:37]: There isn't any requirement for this command to be used in draft PRs
```suggestion
3. Go to GitHub and open up a pull request for this repository, then post a comment on the PR with the text `@metamask-bot preview-build`. (This triggers the `publish-preview` GitHub action.)
```

File: docs/contributing.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+   Commit and push the branch.
+
+3. **Submit a pull request for the release branch so that it can be reviewed and tested.**
+
+   Make sure the title of the pull request follows the pattern "Release \<new version\>".
+
+   If changes are made to the base branch, the release branch will need to be updated with these changes and review/QA will need to restart again. As such, it's probably best to avoid merging other PRs into the base branch while review is underway.
+
+4. **"Squash & Merge" the release.**
```
Comment (Gudahtt) [inline docs/contributing.md:107]: Nit: Maybe worth advising that we double-check that the commit summary is correct here as well?

File: docs/contributing.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+   This step triggers the [`publish-release` GitHub action](https://github.com/MetaMask/action-publish-release) workflow to tag the final release commit and publish the release on GitHub.
+
+5. **Publish the release on NPM.**
+
+   The `publish-release` GitHub Action workflow runs the `publish-npm` job, which publishes relevant packages to NPM. It requires approval from the [`npm-publishers`](https://github.com/orgs/MetaMask/teams/npm-publishers) team to complete. If you're not on the team, ask a member to approve it for you; otherwise, approve the job.
+
+   Once the `publish-npm` job has finished, [check NPM](https://npms.io/search?q=scope%3Ametamask) to verify that all relevant packages has been published.
```
Comment (Gudahtt) [inline docs/contributing.md:115]: Good idea!

File: docs/contributing.md (hunk 2 above)
Comment (Gudahtt) [reply docs/contributing.md:22]: The two last suggestions you listed here seem fine to me! Here are some additional ideas:
 
 * "Using under-development controllers in other projects"
 * "Using controllers in other projects during development/testing"
 * "Testing controllers in other projects"
 * "Testing & local development using controllers in other packages"

File: docs/contributing.md (hunk 7 above)
Comment (mcmire) [reply docs/contributing.md:37]: Ah, sorry. Right. I don't know why I got that mixed up.

File: docs/contributing.md (hunk 2 above)
Comment (mcmire) [reply docs/contributing.md:22]: Okay, I like "Using controllers in other projects during development/testing", I think that's the most accessible version.

File: docs/contributing.md (hunk 6 above)
Comment (mcmire) [reply docs/contributing.md:54]: > it should be easy enough for users to see what the preview build numbers are (it's in the CLI output, and users can always check the package manifests for reference)

It's technically in the CLI output, but it's not very clear IMO. I guess we can improve this developer experience in another PR though.

File: docs/contributing.md (hunk 8 above)
Comment (mcmire) [reply docs/contributing.md:104]: Sure! Added in [bae8925](https://github.com/MetaMask/controllers/pull/993/commits/bae89258a3e9936fa0f6b3c4214a4f011cadd749).

File: docs/contributing.md (hunk 7 above)
Comment (mcmire) [reply docs/contributing.md]: Fixed in [bae8925](https://github.com/MetaMask/controllers/pull/993/commits/bae89258a3e9936fa0f6b3c4214a4f011cadd749).

File: docs/contributing.md (hunk 6 above)
Comment (mcmire) [reply docs/contributing.md]: Updated in [bae8925](https://github.com/MetaMask/controllers/pull/993/commits/bae89258a3e9936fa0f6b3c4214a4f011cadd749).

File: docs/contributing.md (hunk 2 above)
Comment (mcmire) [reply docs/contributing.md]: Renamed in [bae8925](https://github.com/MetaMask/controllers/pull/993/commits/bae89258a3e9936fa0f6b3c4214a4f011cadd749).

File: docs/contributing.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+1. Navigate to your settings within GitHub and [create a **classic** access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#creating-a-personal-access-token-classic). Make sure to give this token the `packages:read` scope.
+2. Switch to your project locally and add a `.npmrc` file with the following content, filling in the appropriate areas:
+   ```
+   @<your GitHub username>:registry=https://npm.pkg.github.com
+   //npm.pkg.github.com/:_authToken=<your personal access token>
+   ```
+   Make sure not to commit this file.
+3. Open the `package.json` for each package that you want to publish and change the scope in the name from `@metamask` to `@<your GitHub username>`.
+4. Switch to your fork of this repository locally and run `yarn prepare-preview-builds $(git rev-parse --short HEAD) && yarn build && yarn publish-previews` to generate preview versions for all packages based on the current branch and publish them to GitHub Package Registry. Take note of the version that is published; it should look like `1.2.3-e2df9b4` instead of `1.2.3`.
```
Comment (Gudahtt) [inline docs/contributing.md:52]: Nit: Bash lint issue, always double quote to prevent string splitting (not sure it's possible for this command to return a space, but, :shrug: )

```suggestion
4. Switch to your fork of this repository locally and run `yarn prepare-preview-builds "$(git rev-parse --short HEAD)" && yarn build && yarn publish-previews` to generate preview versions for all packages based on the current branch and publish them to GitHub Package Registry. Take note of the version that is published; it should look like `1.2.3-e2df9b4` instead of `1.2.3`.
```

File: docs/contributing.md (hunk 6 above)
Comment (Gudahtt) [reply docs/contributing.md]: Agreed, definitely could be made more user friendly

File: docs/contributing.md (hunk 10 above)
Comment (mcmire) [reply docs/contributing.md:52]: Yup, good point.

**Review by Gudahtt — APPROVED** [review]
LGTM, great work!

**Review by legobeat — APPROVED** [review]
